### ISSUE TITLE
Shards/pane track - lots of boosts coded, what comes next? 

### ISSUE THREAD

**Issue by Calamitizer** [eternaldensity/Sandcastle-Builder#1346]
So, I've coded most of what's in my head for the shards/panes track, and most of it is pushed. I'll be backing off of coding new content now as we evaluate what to do next and which boosts need work or need scrapped.

What's left relevant to those boosts is A) unlock conditions, specifically the numbers involved (e.g. how many CatPix does the player need to drain before boost X unlocks, &c.), and B) prices, specifically related to non-shard pacing (e.g. how does this mesh with the bonemeal/vacuum/logicat/qq track). I need a lot of help calibrating both of these things as they're very important for QA. What is SOP for deciding things like this? 

Next question: should this update wait to be released with TaTPix, or not? Personally, I am torn. On one hand, only three boosts (Aperture Science, Dimension Keyhole, and Green Sun, none of which is fully coded since neither is TaTPix) directly relate to TaTPix so far, since I have been focused on handling CatPix. There is plenty of CatPix content to be released, imo. But I also don't want that content to run out of steam before TaTPix is coded, or the whole point of shards is ruined. I would lean to releasing at least some of it earlier, but I'm not sure it would feel like it had a point. What do you guys think? 

In short, I'm just not sure what needs to happen next. I would really appreciate thoughts on the current content of the track, as I've been working mostly in isolation and I don't want to code a bunch of stuff that's considered bad. If I have made any questionable design decisions, please let me know; that's my big question here. I'm still extremely new to both JS and game development so I'm constantly worried about doing something dumb.

**Comment by eternaldensity** [comment]
Putting some CatPix stuff out before TaTPix is done does seem logical to me.

**Comment by pickten** [comment]
For the CatPix time question, we might consider doing something to match the post-Time Time content (not TaTpix, the GLR 5-panel sequence that either is or used to be rotated in varying pattens). It's one of the weirder parts of Time, and nothing relates to it ingame, but wouldn't make sense as a Signpost thing since there's nothing happening besides watching the waves. Since the CatPix show up instead of those, it might make sense to have patterns that match up with Post-Time somehow.

The other thing is that I'm guessing requesting images for TaTPix is pretty easy (at least, I hope it is), and I think I know how to handle the rest of TaTPix implementation, so I can get on that when I get the chance.

**Comment by Calamitizer** [comment]
Sorry, I'm not sure what question you're answering in your first paragraph? What are you proposing we do with that sequence?

By requesting images do you mean requesting permission to use them, or having the game request them from the server? As far as implementation, as far as I'm aware none of it is done. What are you looking at that suggests otherwise?

Regarding the plan that I have in mind, at some (early) point along the shard track, you unlock Aperture Science, which allows you to get Dimensional Keyholes. Buying a Keyhole boosts AS's power, starting at 1, and you can go as far into TaTPix as that number (e.g. buy AS, you can get to TaT1, buy 3 keyholes, you can get to TaT 4). A keyhole always costs 1 pane, which has an increasingly high shard cost. This adds some longevity to the earlier parts of TaT, but pane production will also be steadily boosted, so it isn't just a O(sqrt(n)) grind, and eventually progression becomes smooth. If we go with this plan, there needs to be something in the ONG code that makes TaT N ONG back to itself if AS' power isn't sufficiently high (similar to how NP0 behaves now).

I still think your idea about storing, say, TaT 18 as NP 18.1 is a great one, although there will be lots of places in the code to change to make sure their behavior doesn't change. That'll be something I can definitely help with, but as far as actually implementing TaT itself, I think that's pretty far beyond my abilities.

The more I think about it, the more I'm convinced that TaT should have at least a couple new Stuffs to go along with it, so the more releasing shards sooner rather than later sounds like a good idea. I'll think about where a good place to draw the content line would be.

**Comment by pickten** [comment]
Sorry, I misunderstood what you meant originally, and thought TaTpix couldn't be loaded (the code for which seems to be added already), rather than that the interactions weren't fully done. I'm not fully sure what spots will need to be changed for that, and will go through those soon.

To go with TaT, I agree new Stuffs would be nice. However, I don't feel like there's much in TaTPix that we could use for Stuffs in the same way that Sand/Castles were obvious choices in Time. Any thoughts?

**Comment by eternaldensity** [comment]
There[s some boast, and radioactive waste (blueness).

**Comment by pickten** [comment]
It would be a little weird to have boats for TaT but not Time, so I think blueness is probably going to be the more interesting one to start with. Here are some random ideas on one way Blueness could work. Most of it is probably useless due to the amount of Stuffs and relative lack of content for several, but there might be some usable parts:

Blueness can be collected from TaTpix by clicking (not on Time). Every mNP, about .1% of your Blueness is destroyed (so half-life is 1NP). When Blueness is destroyed, you get 1/3 of that amount of Redness, and 1/3 of the Blueness loss as Greenness. Redness and Greenness are not otherwise obtainable. Redness and Greenness decay similarly. However, you can combine Blueness and Redness to create Purpleness, Redness and Greenness to create Yellowness, and Greenness and Blueness to create Cyanness. Purpleness, Yellowness, and Greenness do not decay. Furthermore, you can combine all three into Blackness (i.e. ink). 

With enough Blackness, you unlock "Polarizer". Polarizer lets you spend lots of Blackness and Cyan/Purp/Yell to create Red/Green/Blue. It also lets you spend even more Blackness to turn Blackness into Whiteness (i.e. erasers), but we'll get there in a minute. Accumulating lots of Redness unlocks "Crimson", which is a prereq for some boosts to Dragons. Accumulating lots of Greenness unlocks "Jungle Green", which is a prereq for some other boosts to Dragons. Accumulating lots of Blueness unlocks "Ocean Blue", which is a prereq for a third set of boosts for Dragons. However, Crimson also enables a significant production of Redness, and Jungle a decent production of Greenness. Ocean also produces a small flow of Blueness. These three allow players to reasonably obtain some Whiteness and make some progress on Dragons.

In the presence of a Blackness, Whiteness and Blackness quickly react and form nothing (i.e. they cancel each other out). Nevertheless, players can easily obtain enough Whiteness to unlock the "Robotic Inker", which allows for automatic crafting of any colors. With some careful balancing, they can also obtain enough for "Not a Priest", which collects a small amount of Grayness when Whiteness and Blackness cancel out. Sufficient amounts of Grayness lets you unlock the "Equilibrium Constant", which can be toggled. When on, Grayness also slowly reacts to form lots of Whiteness and Blackness, so that Whiteness can be accumulated. Sufficient amounts of Whiteness unlocks "Hallowed Ground", which is required for a fourth and final set of color-related Dragon boosts (including one that could allow players to set a Biome before fledging that would affect the locals and cause the Dragons to produce different Stuff/different amounts of Stuff). It also releases a flow (equal to that of Crimson) of Whiteness.

This new flow enables players to get enough Whiteness for "Photoelectricity," which lets "Not a Priest" be disabled and does some useful stuff when "Not a Priest" is disabled (not sure what, but probably a big boost to Redness/Greenness/Blueness/Whiteness production).

**Comment by pickten** [comment]
OK, @Calamitizer, how should people get to TaTpix (code-wise)? Signpost lets you go to 0 and grants 3d lens and stuff, but doesn't seem to enable TaTpix reaching. I'm working on the TaTpix enabling, and I'd like to also test and make sure whatever boost that is works at the same time.

For my own notekeeping and for the curious. Please let me know if anything is missing from both lists. I'll edit as I go:

**Stuff I've done:**

Lots of Math.floor for things that need to be integers. I think I got all of them.

Some helper funcs to make traversing nps easier.

NPdata is now an object (for noninteger keys) and saves without TaT dragon info still work.

**Stuff I know needs to be done:**

Make sure ONG handles nonints. It probably does, but I need to check.

More things need to traverse nps than NPdata management. I'm still looking.

Discoveries/periods/eras need to be coded for TaTpix and be checked.

Make NPdata only store TaTpix within the TaTpix dragon range instead of dragon range

Make Dragon combat data for TaTpix

HighestNP only works for Time. I'll add LargestNP for cross-timeline versions, redefine Highest with get/set to look at LargestNP[0], and adjust accordingly.

Possibly other things need to be ints. I don't think so, though.

Some things to make log 3.1 not the same as log 3 (I wound up using floor for this part. I don't want to have to make this an object as well). I might not do this in the end, though, because I see no reason to and it's potentially tricky. Besides, I already have quite a bit to do.

**Comment by Calamitizer** [comment]
Mmk. First things first, I just now saw your post about Blueness, and I think there's some awesome potential there. I'll table thinking about it for now though, since I think 3.7 is close-ish to release (since we apparently aren't waiting on TaTPix, I guess) and I want to wrap up pricing & playtesting. I can't see anything missing from your list at a glance, but it's a big undertaking that I haven't given much direct thought yet.

As far as TaTPix goes, that's a great list of stuff to do! My thought is that Aperture Science is a prereq for Controlled Hysteresis, which will let you choose where ONGing on NP0 takes you (nowhere, OTC 1, or t1i 1). At first this will be the only way to switch between OTC and t1i, making it a little bit of a commitment where you go (although only an hour-long one). Then, once you arrive at t1i 1, ONGs will take you back to t1i 1 until you buy the first Dimension Keyhole, at which point you can ONG to t1i 2. That ONGs back to 2, until you buy the next Keyhole, &c. Does that sound reasonable?

**Comment by pickten** [comment]
OK, but it'll be some work. I'll add that in for things to do. Blueness was definitely not for 3.7 - just for future expansion.

**Comment by EPSIL0N** [comment]
I'm been away so long and suddenly everything is over my head but good to see we are still active.
Is there content that needs testing by end game players?

**Comment by pickten** [comment]
Oh hey, welcome back! I don't know when you last were active (February? If so, I don't think you missed a whole lot), but there's the following unreleased content: 

Some stuff from Luminous that I'm unsure about (isn't on github yet, so ask Luminous if you want to test that), 

Some dimensional shard/pane stuff from Calamitizer (two stuffs which will be added on release). It probably could use testing, but it's also inaccessible even with the master branch, so IDK. 

There are some minor tweaks, but they were easy enough to test and appear to work. 

I'm working to add the TaTpix for the shard/pane stuff (as in the Signpost proposal a while back) to be able to be finished (eventually), so besides the Dimensional Keyhole, which should work fine, I don't think there'll be any requirements for testing that, which is convenient. If you want to test, go ahead though. I edit purely on github, so the TaTplement branch of pickten/Sandcastle-Builder is always up to date on whatever stuff I'm changing to allow for TaTpix, though it's rarely stable. When it approaches stability (also visible with the list I've been updating above), I'll also push regularly to my gh-pages, so pickten.github.io/Sandcastle-Builder will have that version as well.

The whole blueness thing is a possible thing for the future of TaTpix, so not nearly testable. Besides that, I don't know of any seriously open ideas/currently being coded content. If you check back after 3.7 (the next update, which I think will be soon), we'll definitely have stuff needing testing though, like whatever TaTpix content we add.

**Comment by Calamitizer** [comment]
Hey there! There is *absolutely* endgame content that needs tested, if you would be so kind. Currently there is a ~20-boost track (Shards/Panes) planned for 3.7. A brief overview is that ONGing past The End with all discoveries will unlock Signpost, which will take you to NP zero. ONGing there unlocks 3D Lens, which allows you to snap the camera on CatPix for Dimension Shards. Then there are lots of boosts involving those. I would love to know if something is too easy/hard or takes too long/short or is overly obtuse, since those get harder to assess when you're designing the content. **The primary thing holding back release is pricing/playtesting**, so help with those (especially pricing) would be wonderful. Specifically, I have no idea what non-shard/pane prices things should have (in terms of QQs, Vacuums, and Bonemeal, probably), so as much input as you have on that would be awesome.

Glad to see you back here again =]

**Comment by eternaldensity** [comment]
> Google Analytics has a few leftover highests, because I don't know what should be there. ED, thoughts? You're the only one I know of who sees that, so I figured you might know what's most useful for you.

Hmmm.  looking at the code, it says
```
_gaq.push(['_trackEvent', 'Molpy Down', 'Complete', '' + Molpy.highestNPvisited]);
```
We really need to send it the currentStory and the largestNPvisited for the current story.  Which means either doing it as two separate parameters (i.e. replacing 'complete' with the currentstory value (as a string)) OR by concatenating the currentstory and largestNP strings together as the last parameter.

But I think the former suggestion would work better for reporting since then I can filter by story.





> Blueness can be collected from TaTpix by clicking (not on Time). Every mNP, about .1% of your Blueness is destroyed (so half-life is 1NP). When Blueness is destroyed, you get 1/3 of that amount of Redness, and 1/3 of the Blueness loss as Greenness. Redness and Greenness are not otherwise obtainable. Redness and Greenness decay similarly. However, you can combine Blueness and Redness to create Purpleness, Redness and Greenness to create Yellowness, and Greenness and Blueness to create Cyanness. Purpleness, Yellowness, and Greenness do not decay. Furthermore, you can combine all three into Blackness (i.e. ink).

I'm not sure how I feel about *so many* colours.  Especially since the Blueness is the one that actually appears in any of the TaTpix (particularly the SilentPix, i.e. those by SilentTimer).  That lead to things such as Bluenhancement in the OTT.

On the other hand, the decay and crafting idea does sound really cool.



> I've been away so long and suddenly everything is over my head but good to see we are still active.
> Is there content that needs testing by end game players?

Woo, welcome forward EPSIL0N!  yeah, it is as the others have said.

**Comment by pickten** [comment]
Shards/panes are accessible? If so, oops. Also, Ex-dimensional should have a different unlock, because it does unlock and is annoying to see when I'm nowhere near reaching CatPix and thus nowhere near being able to buy it. Maybe require one of the Stuffs to be owned or CatPix to have been reached (you can use Molpy.largestNPvisited[0] to test that)

Anyways, that was the big problem with the color thoughts. If there were more separate content for each or less of them, it would be better (though still not great), but as is, it's more to have some mechanics that might work. It would be possible to cut Red/Green/Yellow/Cyan/Purple and just have Blueness/Blackness/Whiteness/Grayness (blackness appears, since it's more or less ink, and whiteness is more or less eraser, while grayness sort of gets identified as oxygen by "Not a Priest", and therefore is a big question in t1i with all the discussion of "could one breathe on a mountain?"), but that could get a little weird. 

Decay is definitely something that belongs with the Blueness track in one form or another, for the simple reason that it's radioactive material, and the crafting was sort of fusion-inspired. It would be nice to have some sort of craftable stuff later, though. White/black/gray weirdness was, as noted, an equilibrium reaction, so the original idea felt nice and chemistry-based. "Photoelectricity" was based on the photoelectric effect, for which Einstein won the Nobel Prize, so that would have something to do with mass-energy equivalence, which would make sense.

With the analytics, HighestNP is always LargestNP[current story], due to a get/set I added, so that's what's currently being sent. It'll have the frac in it for newpix tracking reasons (I don't use currentStory often in order to lessen the amount of stuff I need to change), so you'll also get the current story if you look at the decimal part. It's not great though, since that might require looking up the frac->strory chart (which is simple ATM, but won't be as simple if/when more stories are added). The main question was whether you'd want to see all of largestNP instead (i.e. see players' progress in all stories).

**Comment by Calamitizer** [comment]
Oh - yeah, shards are accessible! Sorry, I hadn't realized what you meant earlier. And yeah, oops, I left Keyholes on testing logic levels, pushed a fix. Keyholes are part of TaTPix anyway, and shouldn't be unlocking at all.

Letting my mind wander towards blueness for a bit... Yeah, I had the thought of cutting Yellow/Cyan/Purple too, since they seem sort of extraneous. I wondered about cutting Red/Green too, since they aren't in the comic... but they're interesting, I think. 6 colors is still a lot though, 4 sounds more reasonable (as long as colors take the spotlight of early t1i). Decay sounds like an awesome concept to write content for, very novel.

**Comment by pickten** [comment]
Red/green can be simulated easily. Since they add a lot to stuff to manage, I think it would better to cut them and simulate. Plus, it'll be interesting for the players. Also, Controlled Hysteresis will probably be a tempboost unlocked on ONG with Aperture Science of duration 1.001NP, if that's ok.

**Comment by eternaldensity** [comment]
I like the idea of five colours: white, blue, black, red, green.  (Yellow's
already taken by Mustard anyway.)

But if we did those five colours we might get in trouble with Wizards of
the Coast :P

**Comment by pickten** [comment]
If we did that, how would white/black interact/be obtained? Would grayness exist? How would players overcome rgb decay? Originally, that's what yellow/cyan/purple mainly did: they gave players a way to (expensively) prevent some rgb from decaying. If there's no white/black interaction, what's the difference between white and a lot of black? If there is, how do players overcome/mitigate it? (sorry for the amount of questions, I'm just not sure how that would fit together)

Anyways, five colors (really only 3 to begin with, since black/white come in later) does seem like a bit much to handle, especially given that there's already OTContent unlocked at the same time (roughly) as the TaT content.

**Comment by eternaldensity** [comment]
Good points.

Actually, maybe blueness and colour decay is our chance to finally include
complex numbers!

**Comment by pickten** [comment]
How so?

**Comment by eternaldensity** [comment]
Hmm... perhaps... umm...

No, I can't think of anything that works.

**Comment by pickten** [comment]
Nevermind on the stories, found it on the site. I'm in the middle of the overview edits, so that's next on my list.

**Comment by Calamitizer** [comment]
Well, square roots show up in linear superpositions... and radioactivity could be handwaved to be vaguely quantum, thematically. I will think about it soon, I think there is potential there.

Meanwhile, I *think* every boost for 3.7 is coded, priced, and unlockable as of #1355. Input on the numbers would be really appreciated, but I'm not sure there is much more I can do. Since this content is designed to last a good while, playtesting is very slow. Basically, excepting more input on pricing and an actual picture for NP0, the update is ready, as far as I can tell.

**Comment by eternaldensity** [comment]
there was a suggestion a long time ago about making something involving
complex number arithmetic.  Like some process which rotates a value in the
complex plane, and some things which have imaginary costs, or negative
costs.  Or both.  And there's ways to change the rotation speed, which is
important to get the timing to line up with various processes.  But I don't
think it maps well to the multiple colours idea.

**Comment by pickten** [comment]
I believe Mario is sorted except for a bug I just noticed where vault finding doesn't properly check for Robotic Shopper's being enabled. But since that's at high IP only, and doesn't prevent other things from being found (I think), I'm not 100% sure it's a problem. Regardless, I'll fix it as soon as I finish up the dragon overview changes (very close, I think, just need to make sure it's grabbing the right nps) for TaTpix, but the testing I've done shows Vaults can hit the G range and the tempboosts can be obtained, and I got some of the new boosts from Mario while testing, which would suggest it's otherwise properly done.

NP0 doesn't have a pic in the TaTpix coding (It spits out a blank panel and an error 404 in the console). TaTpix works as far as I can tell on non-draconic stuff (I've yet to double check that all keyholes can be obtained, but I believe I've tested enough that the rest must work) except the options menu (Since my version has lots of changes, could someone test and see if that bug exists in a non-TaT branch? To test, just load the game, load Beachball, and tell me if the options menu gets totally screwed up).

**Comment by Calamitizer** [comment]
Oh, the keyholes are all one boost! Just with a randomized prefix. It should always unlock the same way.

I have been experimenting with Beachball in the current trunk and it seems okay. Is it a visual bug? If so, mine is fine.

Yay Mario. I have calibrated some shard boosts to require about 1e21 locked vaults at once, is that a feasible amount under the current system?

**Comment by Calamitizer** [comment]
ED, do you have any advice about the NP0 picture? Either what it should or shouldn't be, or who to contact to actually get it done if we want to make it something more complex than the sea or a blank screen? For a blank screen (black or white), I am afraid players will think they just couldn't read the resource.

Is there anything else you can think of that should happen before release?

**Comment by pickten** [comment]
For the keyholes, I realized that, I was just unsure whether you could theoretically continue forever, since I got the first and second pretty quickly, but the third took quite a while (even with the hacks I use to test that put me at infinite of a lot of things).

Mario works at a level about 1/1000 your vacuums or sqrt(QQ), whichever is smaller. Since vacuums go inf, that's a bit of a non-issue. For QQs, I don't know how high they go, but I think they'll go inf after this patch anyway, nor do I know what they are when that is meant to be got. The bug is visual. I was just wondering, because it happens on the TaT version, but I've not changed any options or drawing methods (except dragon overview, and even then, only slightly so far). The bug makes the options pane have no max height, so that the options pane goes down to the loot shop on classic without beachball, or through the loot shop with it. 

Frankly, I'd like to have TaT in, so that all your boosts work and so that I don't have to change the version number in the new parts of the load/import functions of persist.js (I *really* hate changing persist.js because of the problems that can ensue if something goes wrong). It's really close right now. As I mentioned, I've tested quite a lot and am finding no serious bugs in the completed parts of the changes (i.e. all but dragon overview changes) except for one that seems to happen only if you hack stuff in too quickly (sometimes, on ONG, newpix can go from TaT to (OTC) -1 if you changed the np number less than 1mNP before ONG and didn't change currentStory). Since the bug goes away if you rehack it in slowly, and doesn't seem to occur if you do things normally, I'm not really concerned about it. Also, players who get there won't feel the lack of content for TaTpix if we put it in this update, since there's so much other new content in this update.

Unfortunately, the amount of changes I've made and am making means merging this may be hard, so I also would like this in by 3.7 so as to avoid letting the difficulty involved in merging get any bigger.

Since there's still that (small) bug in Mario, that should be done before 3.7, but I'm not doing that until dragonoverview changes are done, but that's the last needed feature for TaT (there are some niceties, but they're not needed), so that's another good reason to get TaT in.

**Comment by Calamitizer** [comment]
Mmk, I see. I haven't really given any thought to the Keyhole cost... I suppose I will once it's necessary.

I don't think Vacuums should go to infinite yet? If you have infinite, it may be because of the nest lining reordering bug a few versions ago. WIWTMP should have a power of 8 (or 9 with infinite bonemeal, which a few people have). And I don't think QQs will go infinite either, unless you have something planned? The only related thing I've touched is Tesseract, which still blocks QQ generation. I suppose a loop like Upgrade Mario -> Higher SPP -> More logicats -> Higher logicat levels -> More Keyholes -> Higher TaTPix -> ??? -> More QQs -> Higher Mario could be cool though. I guess increasing the logic cost of successive Keyholes is an alright thing to do, because of SPP.

Ohhh, wait. Is QQ infinite because of GDLP, too? Then WIWTMP could reach 10, sending Vacuums infinite. Is that what you mean?

Anyway, I'm fine with waiting on TaTPix! I didn't know it was so close to ready, honestly, but that's awesome. Right now, I guess I will make sure Aperture Science and Keyholes are ready, since I hadn't given much time to them. I think I will go ahead and code Blueness and whatever it decays into, just for the sake of having something, but we can definitely wait on going further since I plan on making the player spend a good chunk of time in CatPix before TaTPix.

Which (small) Mario bug are you talking about, now? Sorry, I'm getting sort of muddled since it's been in the works for a while.

Also, yes, absolutely, fuck persist.js. =P I recently created a new Chrome profile for testing since I added discoveries to test ```Molpy.Marco()``` and my game didn't want to forget them once they didn't exist. I didn't want to screw around with localstorage.

**Comment by pickten** [comment]
I don't have infinite Vacuums, and just realized the Inf I'd seen was a bugged save as you mentioned (but still, being able to hit about 2^512 vacuums means a pretty huge IP). I have no idea if QQ goes infinite. I just suspected they'd go inf after Luminous adds his update (iirc Luminous said a while back it would include more logicat/QQ stuff). 

Anyways, regarding TaT/Mario:

The small bug was the one I mentioned yesterday or the day before, that causes Vaults to be auto-bought at high IP if the RS has them in the "tobuy" list, even if the robotic shopper has those turned off or it's not ashf but RS is set to ASHF only. Honestly, I don't see much harm in it, since deactivating (or not owning) Aleph One will stop the autobuy, as will removing the name from RS. There's also the task of removing the approx option to avoid letting people break their saves, though 3.7 tested saves will break as a result.

Also, TaTpix came together much more quickly than I expected. If you look above, the things I have listed that I haven't done, but ought to do (besides a bit more testing) has shrunk to "shorten saves if possible" "fix the options bug" "make the new overview storyline indicator look nicer, because it's not aligned right", and "finish the Overview changes". The first and third are things that can wait if needbe without causing problems, the second is of a mysterious origin (honestly, since it's not fully gamebreaking, I'd rather push the rest first, then update my branch, then go to work on it, so that I'm working with more up-to-date files), and the last is what I'm doing now.

Then, there's two things that can go either way that we might want to decide on. Rifting currently doesn't go to TaTpix, under any circumstances (including being at TaTpix to begin with). I like this, personally, because it adds a bit more challenge, but I can probably change it without too much work. Also, Google Analytics gets sent highestNP (which is a mix of the storyline and maxNP), but not largestNP (highestNP for all storylines). This can easily be changed.

Actually, my issue with persist.js isn't as much the fact that I want the localstorage save (the one used for testing) to not totally break because I changed persist (somehow this hasn't happened, except for my warning about importing a 3.7 save into 3.7 with TaT, namely that you need to manually fix largestNP if you do), but more that I really hope no one's save breaks because I didn't change persist exactly right.

Also, speaking of Marco, more discoveries are coming! I've added 76 t1i discoveries (the 55 I listed before, plus 21 more), though more may be needed. I forget the period/era/eon counts, but they were around 25/10/3. Oh yeah, and they work properly. I totally didn't add them just to avoid doing actual work like bugfixing/adding largestNPvisited, I swear.

**Comment by LuminousLeopards** [comment]
Take a save, delete the first character, and import it. You'll end up at NP0 :) It's actually kind of fun to play through that way.

For QQs going inf, that's planned to occur at Dragons, the stage after Wyverns. There will be a boost called the ExistentiSphinx which will give you what is basically a Logicat puzzle on steroids.

Before release, we should play test what we have. I have merged ED's master fork with https://github.com/LuminousLeopards/Sandcastle-Builder-tree-gh-pages and at some point soon will get castle.hobbitcakes.com pointed at it for a live web version. I'll let you guys know when that's ready, and hopefully we can come up with a list of things to test.

I could foresee difficulties merging boosts and badges. I've been keeping all my code local which is probably not best practice; in the future I'll push commits and synch more regularly.

**Comment by eternaldensity** [comment]
> ED, do you have any advice about the NP0 picture? Either what it should or shouldn't be, or who to contact to actually get it done if we want to make it something more complex than the sea or a blank screen? For a blank screen (black or white), I am afraid players will think they just couldn't read the resource.

I'll ask on the OTT.

**Comment by Calamitizer** [comment]
Hey Luminous, good to see you here! 

So, boosts and badges... The game may lose track of which you do or don't have, but if the total number isn't fewer, will it corrupt the save? (Like the dreaded "cannot set property .earned of undefined" error that kept me up for a few hours the other night.) 

Ah, that Mario bug... I do think it's probably something to take care of. It would be fairly obtuse to the player why Rob just doesn't work as stated. I can take a look at it if you want, since you have a lot to do.

**Comment by pickten** [comment]
Sure.

Also, I'm pretty sure your save will work fine if the number of boosts is less, though it will lose track as you mentioned.

**Comment by Calamitizer** [comment]
Maybe it works differently for boosts, but when I made those bogus discoveries and deleted them, my file wouldn't properly load. The error I mentioned happens and the startup code is aborted halfway through, result in... Brokenness.

**Comment by pickten** [comment]
sorry, meant if the number of saved boosts is less than are coded.

**Comment by LuminousLeopards** [comment]
>So, boosts and badges... The game may lose track of which you do or don't have, but if the total number isn't fewer, will it corrupt the save?

The game internally references boosts by ID. If the number is the same or greater but the order is off from that in the save, it will probably load and then bug out later when it finds a reference to something that doesn't exist.

Biggest problem I see is to our own playtesting saves; as long as the order is consistent once it's pushed to the live branch, nobody else should have any trouble.

**Comment by Calamitizer** [comment]
Right, right. I suppose I will just start over from my chirpingmustard save and console my way to where I am now in my playtest, once hobbitcakes is online. Seems simplest. It should work, right?

**Comment by Calamitizer** [comment]
Also, I have fixed that Mario bug.

**Comment by Calamitizer** [comment]
However, the number of vaults opened is still wonky at low power. For example, at Mario power 2, the locks still go down 1 at a time.

**Comment by eternaldensity** [comment]
NP 0 sugestion:

http://xkcd.mscha.org/tmp/wonder.png

**Comment by pickten** [comment]
This is really good, but unusable, for the obvious reason that it's 2 images. Is this an OTC frame? If so, do you know which?

**Comment by eternaldensity** [comment]
> 2 images

It's 1 png with transparancy.

It's not based on a real frame, because that castle never existed on the raft.

It's a refrence to https://xkcd.com/1/

**Comment by pickten** [comment]
By 2 images, I meant it has both the frame and thumbnail (or whatever it's called) in the same png.

Also, geeze it's been a whlie since I read early xkcds.

**Comment by eternaldensity** [comment]
No, it's the second frame, after which they'd floated further away.  Like comic 1.

**Comment by Calamitizer** [comment]
If we use that, I think we should maybe add a boost with a link to that comic, because I'm not sure how many people would get it? I didn't ^^ I do like it now that I know, though.

**Comment by pickten** [comment]
I like it as well. I'll probably add it in post-merge. It just does seem weird having two panels for one frame. Would it work to split the frame into the two panels (stored on the github), then rework the flipping thing so it switches frames at NP0 instead?

**Comment by eternaldensity** [comment]
I don't know what will visually work best, you'll have to try out a few
things and see how it looks.

**Comment by pickten** [comment]
I've created a gist with some updated, more balanced, and more fleshed out thoughts on Blueness to match the discussion here. So far, I balanced (roughly) the stuff we had, cut all but Blue/Gray/White/Black, merged and re-added Red/Green as Otherness, added Photoelectricity (an IP-like boost that really boosts production as a result of what's on it), added Retroactivity (which allows players to purchase specific recurring Photoelectricity boosts, though at a steeper cost) and added a couple supplementary boosts to let production be reasonable. 

Link: https://gist.github.com/pickten/b80d29772174f9f5e2a6

I'm drafting these ideas over there now instead of here, because it's mostly lost in this thread.

I've also started working on the Variegated (the name for the tag for Blueness, etc.) parts in another gist (editing boosts.js directly causes ridiculous amounts of lag for me because its so long and I edit on github only).

**Comment by Calamitizer** [comment]
That all sounds pretty awesome so far! I'm looking forward to it. Also, I humbly suggest Variegated change to Variegation, to match the noun-hood of the rest of the groups.

Soon I will actually decide on an unlock condition for Aperture Science, and then would you like me to work on some of these boosts? Once you get the colors themselves/their decay/other basic stuff, I can start on some of the later ones.

**Comment by pickten** [comment]
For variegation, sure. Though I like variegate, because I really want to use the pun that comes out of it either way.

Anyways, I've got a gist going (linked on the drafting gist) with all important code (i.e. everything that would goo into boosts.js) and notes about all other changes I need to make (not many), so it's moving at a decent pace. I'd also like to get at least the early part coded myself, if that's ok. 

Part of the issue is that Photoelectricity is a little weird: two thirds of it makes Blueness production grow quadratically, while the other half slows it down dramatically, but gives you good stuff if you use it (a splosion), so I know players get a big boost from it (since splosions can never put base production below 1, and Blue's Hints power never decreases), but don't know where they end up at all. The weirdest thing is that, while splosion sounds atrocious, you get a lot of stuff from it (though the amount probably needs to be raised), and chaining resets the timer (though the timer will be very short), so you can get a lot of resources before you take the hit if you get lucky. All in all, all I know is that photoelectricity will speed things up (at least, once I tweak it a bit), and will almost certainly be worth the resources invested. 

So, once I code what I've got, test/balance it (to see where players end up after the dust settles), and add a few more things to even out the beginning (like a timer-increaser for splosion), go ahead and add/code stuff. I'll edit the drafting gist at that point.

**Comment by Calamitizer** [comment]
Okay, sure! I agree, all the early stuff should be done by one person. Let me know how it goes. I can't seem to subscribe to the gist the same way I'm subscribed to this repo, so I'll periodically check.

**Comment by pickten** [comment]
I believe starring does the trick, but I'm not sure. Besides, I edit rather than post comments, and I don't think github has an edits-tracker. This is partly why I've never used gist before. 

Anyways, I've been keeping a small log of what's going on at the top, which is reasonably up to date. I'm making a rigorous crafting system, though, which should be easy to reuse if crafting ever makes a return. The only parts that require new changes, expanding-wise, are Molpy.Crafts, and a function that displays the buttons in the relevant description. I believe the rest is (or will be when complete) automated, including such things as catalysts and multicrafting. 

It's not ideal, though. I've obviously not found everything wrong with it, but:

- The default on the recipe's .times (number of times something can be crafted) actually loops through 1, 10, 100, ... and tries all of them until it finds a multiple it can't craft, which gets a little slow by the time you start getting big numbers. Unfortunately, this is more or less needed by the system, because I needed support for functions as inputs to the crafting requirements/output. 

- There's very little automation in making recipes, because the prototype would already take all the info as input and not really add anything.

- Also, everything runs on boost power. You can't make a recipe that requires boost level, or even one with a specialty unlock, though you can always add that to the filter you use when displaying the result of Molpy.getCrafts (which takes the table and, optionally, the table's power/level, and spits out the list of recipes available under those circumstances. This actually does let you use whatever you want, since the crafter and craft-checker don't check the level, but it's not like you really need it). This is because all Stuffs use power to track how much there is of that Stuff, and because I was lazy and didn't want to code a crazy system just for that to work.

- Generally speaking, the whole setup is not efficient for small tables like Argy, which I wound up coding Crafts-free. It might be faster if I coded an automatic way to get the buttons from getCrafts, but I don't exactly have a good reason to do it, and am not sure what it should have as info, since crafts don't have names.

- Crafts are not saved. No data attached to them will be kept after a load. However, they can still use functions just about anywhere, so I don't really care.

**Comment by Calamitizer** [comment]
That sounds really awesome so far, and it's coming together very quickly! Generalized crafting is something to look forward to.

By the way, I think I've finished bugfixing Mario, so you don't need to worry about it. I'm reasonably certain he always opens exactly the right number of boxes (with exactly the right number of locks left), regardless of power. The approximation just happens within the opening of those boxes. Hooray for the power of math! Thanks Euclid. I'll push it after the TaT merge.

**Comment by pickten** [comment]
Nice! Amusingly enough, I'm in the middle of Photoelectricity ATM, so I have to recreate a version of that system. RewardRepeatable will get a few more options, and the repeatable list as well.

**Comment by Calamitizer** [comment]
Ah, I see. Well, in case it helps, here's how I changed the Mario section of `Molpy.RunFastFactory()`:

```javascript
for (var i = 0; i < runs; i++) {
    Molpy.RewardLogicat(Molpy.Level('QQ'), Math.floor(l/runs);
}
Molpy.RewardLogicat(Molpy.Level('QQ'), l % runs);
```

It comes from the identity a = floor(a/b)*b + a%b (division algorithm). This technically does `runs + 1` calls of `Molpy.RewardLogicat()`, but who cares.

**Comment by pickten** [comment]
Why not do runs-- at the beginning? It solves that problem quite easily.

Anyways, makes sense, though I'm going to use the approach I used originally (1/25, 1/24 left, 1/23 left, ...) because it's nice and easy. always spits out the right answer, and keeps all runs roughly the same size (plus or minus 1), which looks nice to the player.

**Comment by eternaldensity** [comment]
:+1:

**Comment by eternaldensity** [comment]
Here's another suggestion for NP0, from mscha:

http://xkcd.mscha.org/tmp/np0.png

It's a bit simpler and more obvious than the other idea.

**Comment by pickten** [comment]
That's pretty neat. It would make a lot of sense as NP0 of OTC, and it works fine for coding purposes. I'll add it.

Also, if you didn't see the note on the drafting gist, I've got the Blueness parts planned earlier as fully coded as I can get before editing SB's files (i.e. all code needed is on a separate file) and before balancing. This goes all the way up through Photoelectricity (it says up through Diluted/Concentrated Boom, but Photoelectricity is a bit mario-like and these are the most recent boosts I've come up with for it to unlock), which is really weird and unpredictable because of the three repeatable boosts: splosion, blue fragment, and Atomic Pump. I've also got a jsfiddle (linked on the drafting gist) that I'm editing to try to simulate Photoelectricity (it's outdated though because infinites were happening, and, besides, is still broken). It's really just that weird. I think it does it's job, though, which is to have splosion produce quite a bit of stuff, to have ocean blue vary around 1-3, and have hints slowly creep up at a reasonable rate. I think this moves players to 10k levels. Really not sure, because I actually had a few (very slow) inf-causing cycles before, so I changed a few things that really affected the balance. Anyways, this may sort of be dependent on having a bit of outside content to smoothen the bumps as photo does its thing. We'll see.

**Comment by Calamitizer** [comment]
Woo! Everything works pretty well so far. I've pushed some minor touchups, and I can successfully ONG into and out of t1i, and successfully advance t1i by buying keyholes. Good job!

So far, the problems I've encountered are
A) `Molpy.highestNPvisited` is set to 40 (?) when I'm in OTC
B) ONGing normally at NP3430 (without signpost or hysteresis) sends me to NP2
C) Some extreme shenanigans ensue when you switch hysteresis and you're not on NP0, but that wasn't intended anyway, so I removed that option.

**Comment by Calamitizer** [comment]
Oh, and D) Using Negator on NP n sends you to NP -(n+1) in t1i. Should be a simple floor/ceil swap but I'm not sure where to put it.

**Comment by eternaldensity** [comment]
The negator button does:
```
onclick="if(Molpy.Spend({FluxCrystals:1})){Molpy.newpixNumber*=-1;Molpy.UpdateBeach()}"
```
It really ought to be moved into a real function rather than text, and any
necessary cleanup can then be added to it.

**Comment by Calamitizer** [comment]
You're right, I'll write that in a couple hours.

**Comment by pickten** [comment]
The first two bugs are strange, since highestNP is always whatever story is being looked at's largestNP value, and ONG has nothing with the number 2 anywhere in it. I'll look at that. One question though: for the first, when is highestNP set to 40? Is it when you arrive, or when ong/mNP happens? And for the second, is it just NP3430?

My guess is those are tied to storyline switches and possibly negative np numbers. It's also probably related to the weird bug that happens (afaik only) if you manually change np and then manually ONG within 1mNP.

Also, I can take a look when I finish transferring blueness (should take 10-20 minutes more).

**Comment by LuminousLeopards** [comment]
I should be ready with my stuff in two to three days. I just need to code Marketing and one other Breath-related boost, do some cleanup and a couple other minor things. Most of my stuff still needs to be playtested, though.

I had a boost named Catalyzer, but it sounds like that might be confusing w/r/t what you guys have coded. I'll think of a different name.

**Comment by pickten** [comment]
Catalyzer should be fine. My stuff has the following boost names (in no order): (stuffs) Blueness, *Other*ness (alias Otherness), Whiteness, Blackness, Grayness, (boosts) Argy, Ocean Blue, Blue's Hints (alias bluhint), Meteor, Hallowed Ground, Polarizer, A Splosion (alias splosion), Blue Fragment, Atomic Pump, Diluted Bomb, Concentrated Bomb, Not a Priest (alias NaP), Equilibrium Constant, and Photoelectricity. 

None of those are very similar, except possibly Equilibrium Constant or Polarizer, but those are still pretty clearly different. You can also see Calamitizer's boosts, and they don't have similar names to Catalyzer either.

The word catalyst was only running around because I added a crafting system, and was explaining its API above, and it allows for catalysts in recipes. Still, players probably aren't going to see that, since crafting itself happens on boosts. They also probably won't know about catalysts in crafting for the simple reason that no recipes currently use any. So, if you think Molpy.Boosts['Catalyzer'] is different enough from Molpy.Crafts[table_name][recipe_number].recipe.catalysts, you're fine.

Anyways, I'll almost certainly be ready with the first stage of Variegated content by the time you are with your content since it's already fully coded and just needs bugfixes/testing, so there'll be that as well. This is shaping up to be a heck of an update.

**Comment by Calamitizer** [comment]
Dare I say 4.0? =3

**Comment by Calamitizer** [comment]
Back home now. I'll investigate that `highestNPvisited` glitch in a bit, I'm having trouble figuring out exactly how to "retry" since I've had problems with localstorage.

**Comment by pickten** [comment]
You're having trouble with localstorage too (there was a reddit post about someone else having trouble with this)? As far as I can tell from the reddit post, it seems Firefox did something weird to it in an update.

Anyways, if we're doing this much for 3.7, I know what one part of 4.0 will have to be. I'm just hoping I have time to finish drafting that part (I've been working on that since late may/early june, because it's the first part of two, it needs to include on the order of 75 boosts, and it wouldn't make much sense to only have part of it). The drafting is about 1/2-5/8 done for the first part (11 pages in google docs, though a lot of it is notes clarifying stuff!), but quite a few of the first 1/4 need to be changed around a bit, I may have to change some parts around to account for 3.7 content, and the back half of drafting this part is harder since it involves content I'm not very familiar with.

And no, I'm not going into any more detail about it now, and probably won't until I'm at least 7/8 done drafting the first part, besides the fact that it probably won't affect the progression in any way that could screw things up. *cue evil laugh*

**Comment by LuminousLeopards** [comment]
I would be good with versioning it as 4.0. There are a couple relatively important additions to dragon stuff as it is, without even considering all the content you guys have contributed. I was tentatively going to go with the version name __EXCELsior!__ since it feels like I've spent more time playing with spreadsheets than actually coding lately.

**Comment by Calamitizer** [comment]
I use Chrome actually; I'm 95% sure that I have brought all my localstorage problems on myself by adding and removing discoveries to test `Molpy.Marco()`. =[

I feel like at this point we need at least three different version names, lol. I'm very partial to "One More Time" for obvious reasons. =p

**Comment by Calamitizer** [comment]
Weird things are happening with Now Where Was I? that I'm not sure how to fix. My `Molpy.highestNPvisited` is 1.1 even when I'm in OTC, resulting in NWWI taking me into t1i.

Also, sometimes it's possible to get the NP info box to say "Newpix 0 of t1i".

**Comment by pickten** [comment]
For versioning, why not 4.0 and "One More EXCELlent Kitty Genocide Simulator" It's a mix of all three names, and one of the only combinations I could figure out that made any sense. Alternatively, "EXCELsior: Simulate Kitty Genocide One More Time" which is longer, but almost has the full names (though KGS had to be swapped around and lose the 2015).

That bug's never happened to me before. Oh yeah, whatever redraws the box needs to do it after ONGs() because Hysteresis just sets NP to the frac part, so that ONGs() can take care of incrementing it. Once I start variegated testing, I'll look into that. Also yeah, where was I was the last change I made and it was a pain because of Molpy.TTT... which might be the source of the bug.

So, anyways, 0.1:0 should happen only at first load/transfer, and fractional np nums should never ever happen on OTC. What did you do to get this to happen?

**Comment by LuminousLeopards** [comment]
4.0 it is. I have no strong feelings about the name; we could take the EXCEL part out and use it later if nothing else. Up to you guys.

**Comment by Calamitizer** [comment]
I would be fine with "One More Time: Kitty Genocide Simulator 2015". EXCELsior is a good name, but I have a feeling as long as dragon content is being developed, there will be no shortage of spreadsheets lol.

I notice now that there are lots of things in `persist.js` under `if (version < 3.7)` that I never ran because I was testing on something called 3.7. Later I will try restarting my save and running those and see whether it fixes those bugs. My contributions will probably taper off somewhat over the next few days, though, since school is starting on Wednesday.

**Comment by pickten** [comment]
Ouch. I think I have an orientation thing in a few days, but nothing else for almost a week after.

Anyways, there were a few versioning things in persist.js that came from the TaT changes. Still, I don't think there were that many changes that the bugs you were mentioning could happen. There were int->float changes, and there was a defaulter for highest/largest, but that was pretty much it, I think. 

Oh, actually, there were dragon changes as well, so dragon stuff probably got somewhat broke for you.

**Comment by eternaldensity** [comment]
From the amount of stuff we've done, I'm okay with going directly to 4.0.

I think I'll even announce the update on /r/incremental_games when we make the update, cos it's significant enough for that.

> Anyways, if we're doing this much for 3.7, I know what one part of 4.0 will have to be. I'm just hoping I have time to finish drafting that part (I've been working on that since late may/early june, because it's the first part of two, it needs to include on the order of 75 boosts, and it wouldn't make much sense to only have part of it). The drafting is about 1/2-5/8 done for the first part (11 pages in google docs, though a lot of it is notes clarifying stuff!), but quite a few of the first 1/4 need to be changed around a bit, I may have to change some parts around to account for 3.7 content, and the back half of drafting this part is harder since it involves content I'm not very familiar with.
> 
> And no, I'm not going into any more detail about it now, and probably won't until I'm at least 7/8 done drafting the first part, besides the fact that it probably won't affect the progression in any way that could screw things up. cue evil laugh

:D :D

**Comment by pickten** [comment]
Well, not 4.0 then. But I'll finish by 5.0.

**Comment by eternaldensity** [comment]
There's a limitless number of numbers between 4.0 and 5.0.  The numbers only go up as quickly as we do stuff.

**Comment by pickten** [comment]
True, but it's not like the rate that the version goes up changes all that much (besides the 3.67->4.0 we're doing). Besides, I only said "by 5.0", not "for 5.0". It can certainly be part of 4.5 if I get it done fast enough, or even earlier.

Also, I said I wouldn't spoil it, but you can probably figure the basic idea out already with a bit of thought from what I've already said, the stuff that has been in the game since before this update, the fact that this update doesn't accomplish any part of it, and this sentence.

**Comment by Calamitizer** [comment]
... Nope, I got nothing. =p

**Comment by eternaldensity** [comment]
Yet again, I managed to hold contradictory thoughts (or forget that 4 equals 4).  Oops.  Yeah, if we're going to 4.0, then the next big thing can be 5.0.

**Comment by eternaldensity** [comment]
Or maybe 2.7 is a big enough change, as the last 26 versions have been 2.6x

In any case, I expect the changelog will require some filling out.  There's a fair amount in the readme but I doubt that's absolutely everything (I'll be happy to be proven wrong)

**Comment by pickten** [comment]
Yeah, the readme doesn't have Variegation (but does have TaT) or Luminous' stuff.

And 4 is totally NaN, so 4!==4 is reasonable.

**Comment by LuminousLeopards** [comment]
Last time I just dug through the Git history and copy+pasted commit descriptions; seemed to work fairly well. I have a log of all my changes as well.

eternaldensity - how exactly did you configure things to get castle.chirpingmustard.com pointed at your GitHub Pages branch? I have a CNAME DNS record pointing sandcastle.hobbitcakes.com to luminousleopards.github.io, and a CNAME file in my repository pointing the other way, but it just gives a GitHub 404.

Alternatively, if you wouldn't mind forking the master again and setting it up under a different subdomain, we could host a test copy of the game on your site. Either way works for me, I just think that for such a substantial release there should be some playtesting so we don't break people's saves or cause mass confusion.

**Comment by pickten** [comment]
You're missing a gh-pages branch in your repo. That's the one that gets used, I'm pretty sure. Anyways, if this doesn't work, we can always use github pages to test, but we should try to have a dedicated url that has the ED master (might take some work) so that we (or curious players) can test what's coming easily.

When we're ready, we can just post our own logs here and put them together.

**Comment by Calamitizer** [comment]
Because of the shenanigans going on with save data / version updating / constant updates I am wary of encouraging non-devs to try trunk. But a game called Dungeon Crawl Stone Soup does a system much like this.

**Comment by pickten** [comment]
True. Maybe we could push to it after every stable update to master, and do smaller version increments (i.e. instead of 3.67->3.7, do 3.67->3.671->3.672->...) in unfinished updates to account for that, (and probably start using <= instead of < in `persist.js` versioning checks). It would be nice to have a dedicated testing system, though. It's also not that uncommon these days for games to have unstable/nightly branches like that (e.g. Stone Soup, Starbound), and players who do that sort of thing usually do at least understand that they need to update a lot more often. 

Anyways, just having a fixed testing URL, even if its only active when we're ready to playtest, would be nice.

**Comment by Calamitizer** [comment]
I definitely agree with at least a plastering branch. I do think a stable trunk would be fine, as long as it includes a warning something like "If you aren't comfortable with JS/console commands/source-diving, don't get to attached to these saves".

**Comment by Calamitizer** [comment]
... *playtesting. Thanks phone.

**Comment by EPSIL0N** [comment]
Yeah, we need to be sure the save situation is somewhat idiot proofed. I liked what we have but if there are so many of us working in parallel I guess it would be good to have a more common branch to push tests to.

**Comment by eternaldensity** [comment]
Yes, gh-pages branch is needed for your github to be online.

https://help.github.com/categories/github-pages-basics/

And the proper url to be pointing to must include the project name
(capitalised correctly) like
http://eternaldensity.github.io/Sandcastle-Builder/

**Comment by EPSIL0N** [comment]
I was curious, do we have anywhere where people are making note of what they are working on or is it mainly just this thread so far? I am starting to play with ideas tho nothing as big as what some of you are working on so far.

**Comment by Calamitizer** [comment]
Not really. I posted a bunch of my thoughts on shards in this topic and a previous topic, and pickten has his gists he was working in, but other than that there's not really anything like that I know of. Feel free to share your ideas if you want, I'd love to hear them!

**Comment by pickten** [comment]
Would it be worth looking into starting some sort of gist/doc/whatever to keep track of what's being worked on (and by whom)?

**Comment by EPSIL0N** [comment]
It might be good but I am not really sure what we need, I was just trying to figure out if anyone is working on any dragon related stuff without having to make another topic that is just asking that as Dragons, Prizes and TaT were among the big topics a while back and I have begun to fuss over the Dragons in my own mind.

**Comment by LuminousLeopards** [comment]
I'd favor Gist for a "live" view of what's being worked on. There's also piratenpad or Google Docs.

For playtesting: https://docs.google.com/spreadsheets/d/1ry0g-n26zvJiuPVQr4AnTlcY8B2F3_6FTAUa5sYloaA/edit#gid=493290078 . The idea is you put every discrete "thing" that you've been working on and possible problems/concerns/etc. and other people can check your work.

I don't have a pretty URL but http://luminousleopards.github.io/Sandcastle-Builder/ hosts a copy of the game that will be suitable for playtesting once I merge in my changes (tonight). You are all collaborators so you can merge pull requests or edit the code itself on GitHub. The repo URL is https://github.com/LuminousLeopards/Sandcastle-Builder-tree-gh-pages

I have a modified-for-Wyverns version of the dragon simulator that I will add into that spreadsheet as well.

**Comment by LuminousLeopards** [comment]
http://luminousleopards.github.io/Sandcastle-Builder/ is live with my changes! I am still working on several things but you should be able to upgrade to Wyverns and start playtesting (pretty pretty please). There are probably several things broken as I had to manually resolve a bunch of conflicts. From now on I will commit a lot more often to avoid this...

My somewhat ugly revision of the dragon simulator spreadsheet is up on Google Docs as well, so please feel free to use it if you're playtesting.

**Comment by eternaldensity** [comment]
Some relevant game exports would ptobably be a good idea too.

**Comment by EPSIL0N** [comment]
I feel stupid, is there a link to the simulator somewhere?

**Comment by LuminousLeopards** [comment]
Yep, it's here: https://docs.google.com/spreadsheets/d/1ry0g-n26zvJiuPVQr4AnTlcY8B2F3_6FTAUa5sYloaA/edit#gid=493290078

There are also several saves (although more are always appreciated).

**Comment by EPSIL0N** [comment]
Well play testing is going strangely for me. 

Bugs:
I can't seem to get Mario to run reliably, possibly not even at all.
My dragon overview has gone black on Firefox. I haven't tried fledging any new Wyverns yet.

Maybe a bug:
I seemed to be missing one discovery and had not made it or it's twin into sand and glass, very strange.

**Comment by pickten** [comment]
The overview's hiding the indicators is a known bug and I'm trying to pin it down. If there's something besides that, though, please let me know.

For mario, I've not had that happen in my testing. What happens if you try to run Mario? Just to be sure, are you sure you don't run out of QQs? Also, could you provide a save?

**Comment by Calamitizer** [comment]
Hmm, where are you playtesting? I realized today that my Mario on Luminous' version doesn't work (I think?) but my personal copy does. Not sure if it's version async or something going wrong. 

Also, I'm realizing now that that missed discovery is probably -0. I haven't looked at the code, but I think MakeQuadBadge() probably ends up making one. I don't know what to do about that; I haven't been able to contribute almost at all since university started again for me.

**Comment by EPSIL0N** [comment]
I believe I am playing from luminous SCB's io page (not the LL SCB-tree-gh repository) where I am having the Mario issue.

I was missing discovery 2302 but I could have been missing it for almost a year now. After I got it signpost appeared.

**Comment by pickten** [comment]
OK, I'll scan the relevant pieces of code and update accordingly.

**Comment by EPSIL0N** [comment]
Another interesting behaviour or two.

Negative setting still appears to show blank but positive NP grid for me.
Switching display type caused the overview frame to grow incredibly tall compared to a standard layout.

**Comment by pickten** [comment]
Aware of the first (there are erroneous math.abs, but it's not like you can fledge on negpix anyways -- I'll fix it, but it's less high-priority than, say, getting the indicators working). I know the source of the second and suspected there'd be something there, but it doesn't make itself shown on classic so I didn't notice.

**Comment by EPSIL0N** [comment]
Hopefully no one else was working on it as I think I found an issue, quite possibly the issue with Mario and Rob and the "unlock repeatable boosts" function. It appears that it is having issues with the scenario where it would be buying for example Vault Keys only when ASHF is active. Mario was working but doing nothing as:

Vault Key is locked and ASHF is inactive.
Because ASHF is inactive the first conditional group for shouldbuy is false, the second is false because that is only true when always buy is set for Rob and the third conditional is not valid for Vault Key. Shouldbuy stays 0 / false and so Vault Key does not unlock.
Vault Key being locked means RewardLogicat will never reach the line where it falls back to RewardRedacted.

I submitted as a pull request in case pickten or another contributor has already worked up a more elegant solution.

**Comment by pickten** [comment]
Ah, good catch.

**Comment by EPSIL0N** [comment]
Did a little testing and it looks like everything is still well contained in the new loops and the boost silence extends to the RewardRedacted if it falls back to that.

**Comment by EPSIL0N** [comment]
So, I am not sure what to do now. I've been squishing bugs and just generally playing. TaT is still a ways out and something I wanted is ~20 days away.

Not entirely sure what to do with Dragons and not sure if we do or do not want a boost to improve QQ yield or one to make it possible to get more than 2*Infinite FC an hour.

How are things going for other players?

**Comment by pickten** [comment]
I've not yet caught up in the game, so I can't really get to any of the new content. :( As a result, when I test I wind up with unlikely/impossible saves which makes balancing harder, so I've been mostly relying on bug reports and a lot of mental math to see how things should play out. Speaking of bug reports, confirm or deny for the overview fix?

**Comment by EPSIL0N** [comment]
As far as I can tell it is still broken on the testing page using my ongoing testing save but I could probably experiment with other saves and a different computer.

**Comment by EPSIL0N** [comment]
Can anyone check if the classic page works in the testing branch, I believe I get an error related to something in the Github save system there before the page even finishes starting.
